File: Python/lld-examples/easy/parking-lot-system/parking_ticket.py
from datetime import datetime
from typing import Optional
import uuid
from vehicle import Vehicle
from parking_spot import ParkingSpot

class ParkingTicket:
    """Parking ticket class"""
    
    def __init__(self, vehicle: Vehicle, parking_spot: ParkingSpot):
        self.ticket_id = str(uuid.uuid4())[:8].upper()
        self.vehicle = vehicle
        self.parking_spot = parking_spot
        self.entry_time = datetime.now()
        self.exit_time: Optional[datetime] = None
        self.is_paid = False
        self.amount_paid = 0.0
    
    def get_formatted_entry_time(self) -> str:
        """Get formatted entry time"""
        return self.entry_time.strftime("%Y-%m-%d %H:%M:%S")
    
    def get_formatted_exit_time(self) -> str:
        """Get formatted exit time"""
        if self.exit_time:
            return self.exit_time.strftime("%Y-%m-%d %H:%M:%S")
        return "Not exited"
    
    def get_parking_duration_in_hours(self) -> int:
        """Get parking duration in hours"""
        end_time = self.exit_time if self.exit_time else datetime.now()
        duration = end_time - self.entry_time
        return max(1, int(duration.total_seconds() / 3600))  # Minimum 1 hour
    
    def mark_as_paid(self, amount: float):
        """Mark ticket as paid"""
        self.is_paid = True
        self.amount_paid = amount
        self.exit_time = datetime.now()
    
    def __str__(self):
        return f"Ticket {self.ticket_id} - {self.vehicle} - Spot {self.parking_spot.spot_number} - {self.get_formatted_entry_time()}" 

File: JAVA/lld-examples/easy/parking-lot-system/src/main/java/com/systemdesign/lld/parking/ParkingTicket.java
package com.systemdesign.lld.parking;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;

/**
 * Parking ticket class
 */
public class ParkingTicket {
    private static int ticketCounter = 1;
    
    private String ticketId;
    private Vehicle vehicle;
    private ParkingSpot parkingSpot;
    private LocalDateTime entryTime;
    private LocalDateTime exitTime;
    private double fee;
    private boolean isPaid;
    
    public ParkingTicket(Vehicle vehicle, ParkingSpot parkingSpot) {
        this.ticketId = "TICKET" + String.format("%06d", ticketCounter++);
        this.vehicle = vehicle;
        this.parkingSpot = parkingSpot;
        this.entryTime = LocalDateTime.now();
        this.exitTime = null;
        this.fee = 0.0;
        this.isPaid = false;
    }
    
    public long getParkingDurationInMinutes() {
        LocalDateTime endTime = exitTime != null ? exitTime : LocalDateTime.now();
        return ChronoUnit.MINUTES.between(entryTime, endTime);
    }
    
    public long getParkingDurationInHours() {
        long minutes = getParkingDurationInMinutes();
        return (minutes + 59) / 60; // Round up to next hour
    }
    
    public void markAsPaid(double fee) {
        this.fee = fee;
        this.isPaid = true;
        this.exitTime = LocalDateTime.now();
    }
    
    public String getFormattedEntryTime() {
        return entryTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
    }
    
    public String getFormattedExitTime() {
        if (exitTime != null) {
            return exitTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
        }
        return "Not exited";
    }
    
    // Getters
    public String getTicketId() {
        return ticketId;
    }
    
    public Vehicle getVehicle() {
        return vehicle;
    }
    
    public ParkingSpot getParkingSpot() {
        return parkingSpot;
    }
    
    public LocalDateTime getEntryTime() {
        return entryTime;
    }
    
    public LocalDateTime getExitTime() {
        return exitTime;
    }
    
    public double getFee() {
        return fee;
    }
    
    public boolean isPaid() {
        return isPaid;
    }
    
    @Override
    public String toString() {
        return "ParkingTicket{" +
                "ticketId='" + ticketId + '\'' +
                ", vehicle=" + vehicle +
                ", parkingSpot=" + parkingSpot +
                ", entryTime=" + getFormattedEntryTime() +
                ", duration=" + getParkingDurationInHours() + "h" +
                ", fee=" + fee +
                ", isPaid=" + isPaid +
                '}';
    }
} 